# Assignment 2

## Instructions - Please read these instructions THOROUGHLY First! 

This is an individual homework assignment. The implication of this is that:

- You may discuss the problems in this assignment with other students in this course and your instructor/TA, but YOUR WORK MUST BE YOUR OWN.
- Do not show other students code or your own work on this assignment.
- You may consult external references, but not actively receive help from individuals not involved in this course.
- Cite all references outside of the course you used, including conversations with other students which were helpful. (This helps us give credit where it is due!). All references must use a commonly accepted reference format, for example, APA or IEEE (or another citation style of your choice). **You may add these as a separate Markdown cell underneath the references that have already been included.**

If any of these rules seem ambiguous, please check with with your instructor for help interpreting them.

We suggest completing this assignment using the provided notebook. Each question should be answered using a SQL query (or combination or SQL queries) unless the text indicates that you may do something else. You may submit your queries embedded in Python, as calls to `pandas.read_sql()`
## When you submit your work

Your submission will be graded manually. To ensure that everything goes smoothly, please follow these instructions to prepare your notebook for submission to the D2L Dropbox for Assignment 2:

- Please remove any print statments used to test your work (this is done by commenting them out)
- Please provide your solutions where asked, do not alter any other parts of this notebook.
- If you need to add cells to test your code please move them to the end of the notebook before submission- or you may include your commented out answers and tests in the cells provided

## Introduction

 In this assignment, we will focus familiarizing you with using SQL for data exploration, and continuing to cultivate a sense of curiosity about the datasets you encounter. We will be using a CSV File generated by the <b>City of Edmonton</b> containing Licensed Pets in the city. It is assumed that this table has been pre-cleaned (although potentially not entirely) so that you can work on the actual assignment more quickly. This assignment has four (4) parts: PARTS A, B, C, and D.
 
 To begin, start by importing the provided CSV into your own SQL database using SQLAlchemy, by filling in the lines below:

In [3]:
import pandas as pd
import sqlalchemy as sq
from sqlalchemy import create_engine,text

In [ ]:
# read in your CSV as a dataframe
pet_df = pd.read_csv('Pet_Licenses_by_Neighbourhood.csv')

# connect to your database; include a cell at the bottom of this notebook to dispose of your engine object
DATABASE_URL = "mysql+pymysql://root:your_password@localhost:3307/pet_liscensed"
engine = create_engine(
    DATABASE_URL
)
# write your dataframe into a table
pet_df.to_sql(
    name = 'liscensed_pets',
    con = engine,
    if_exists= 'replace',
    index= False
)

pet_df2 = pd.read_sql('Select * from pet_liscensed.liscensed_pets', con= engine)

In [5]:
# demonstrate that your import has been successful by reading your database table as a dataframe, 
# and print some information (not the entire table) about your second dataframe
display(pet_df2.head(5))
print(f"number of records: {pet_df2.shape[0]}")
print(f"number of features: {pet_df2.shape[1]}")
print("Na values in each columns:")
display(pet_df2.isna().sum())

# do not forget to dispose of your database objects, at the bottom of this notebook.

YEAR     MONTH  MONTH_NUMBER   LICENSE_DATE    NEIGHBOURHOOD  \
0  2024  November            11  November 2024             None   
1  2025       May             5       May 2025  Beverly Heights   
2  2023   October            10   October 2023     Wîhkwêntôwin   
3  2025    August             8    August 2025         Capilano   
4  2025  February             2  February 2025     Abbottsfield   

  NEIGHBOURHOOD_ID PET_TYPE                  BREED  GENDER SPAYED_OR_NEUTERED  \
0             None      Dog                UNKNOWN  Female                Yes   
1             2100      Dog     NORWEGIAN ELKHOUND  Female                Yes   
2             1151      Dog                UNKNOWN  Female                Yes   
3             6061      Cat  DOMESTIC (SHORT HAIR)  Female                Yes   
4             2010      Dog  AUSTRALIAN CATTLE DOG    Male                Yes   

    LATITUDE   LONGITUDE                                  LOCATION  COUNT  \
0        NaN         NaN                                      None      1   
1  53.564122 -113.402652  (53.56412226688728, -113.40265235200434)      1   
2  53.541983 -113.523994  (53.54198293050736, -113.52399393494247)      1   
3  53.554672 -113.422792  (53.55467208922095, -113.42279236815314)      1   
4  53.574143 -113.388758  (53.57414343327946, -113.38875800528291)      1   

                                  Point Location  
0                                           None  
1  POINT (-113.40265235200434 53.56412226688728)  
2  POINT (-113.52399393494247 53.54198293050736)  
3  POINT (-113.42279236815314 53.55467208922095)  
4  POINT (-113.38875800528291 53.57414343327946)

number of records: 83468
number of features: 15
Na values in each columns:


YEAR                   0
MONTH                  0
MONTH_NUMBER           0
LICENSE_DATE           0
NEIGHBOURHOOD         31
NEIGHBOURHOOD_ID      31
PET_TYPE               2
BREED                 35
GENDER                52
SPAYED_OR_NEUTERED     0
LATITUDE              91
LONGITUDE             91
LOCATION              91
COUNT                  0
Point Location        91
dtype: int64

## Part A: Warm-up Questions (12 marks)

Answer the questions below, including the queries you used where necessary. Not all questions will require writing a SQL query to answer.

**(A1): (1 mark)**

How many records are there in total?

There are 83468 numbers of records 

**(A2): (1.5 mark)**

How many known areas were covered in the dataset? (<b>Please show the SQL query</b>)

In [6]:
def known_area():
    query = """
    select count(*) as total_number 
    from liscensed_pets 
    where NEIGHBOURHOOD is not null;
    """
    return pd.read_sql(sql = query, con = engine)

print(known_area())

   total_number
0         83437


There are 83437 known areas in the datasets

**(A3): (1.5 mark)**

How many unknown areas were covered in the dataset? (<b>Please show the SQL query</b>)

In [7]:
def unknown_area():
     query = """
     select count(*) as total_number 
     from liscensed_pets
     where NEIGHBOURHOOD is null;
     """
     return pd.read_sql(sql = query, con = engine)
print(unknown_area())

   total_number
0            31


There are 31 unknown area in the datasets 

**(A4): (4 marks)**

Explain what each of the following columns is used for. You may use the original page to guide your explanation (but you should cite it)
* LICENSE_DATE (1 mark)
* ANIMAL (1 mark)
* BREED (1 mark)
* COUNT (1 mark)


LICENSE_DATE: Month and year date when the pet is lisenced, tracking the date time of the liscenses

ANIMAL: Categorization of lisenced pets, seperate the different pets into different categories such as: dog, cat, pigeon 

BREED: A group of animals with homogeneous apperances, selected from its original breeding, using to seperate the pet having the same categories. 

COUNT: Unique count for the record, using to track the number of record for different species in different region. 

**(A5): (4 marks)**

The webpage for the dataset (see [References](#References)) identifies some specific limitations for this data. How is this different from the Calgary licensed pets dataset covered in class?

The data set was the updated version for the last update date, thus if a pet was liscensed for years but its lisence has been expired and have not been reliscensed, it will not be recorded, this is the reason why there only value of 1 in the column of record. On the other hand, the Calgarian dataset does still include the one that have not been relisenced. This make it is available for the Calgarian dataset to make an analysing on trend based on years while Edmoton datasets can not do that. 

## Part B: Simple questions (10 marks) 

For these queries, produce a query which provides the answer.

**(B1): (2 marks)**

List in <b>descending order</b> five (5) neighbourhoods with the highest numbers of licensed cats

In [8]:
def top5_cat_region():
    query = """
    SELECT 
        NEIGHBOURHOOD, 
        COUNT(*) as total_cat_count 
    FROM liscensed_pets 
    WHERE PET_TYPE = 'Cat'
    GROUP BY NEIGHBOURHOOD
    ORDER BY COUNT(*) DESC
    LIMIT 5;
    """
    return pd.read_sql(sql = query, con=engine)

print(top5_cat_region())

  NEIGHBOURHOOD  total_cat_count
0  Wîhkwêntôwin              680
1      Downtown              414
2    Strathcona              362
3     Westmount              359
4    Summerside              354


The list of the top 5 region is: Wîhkwêntôwin, Downtown, Strathcona, Westmount, Summerside


**(B2): (2 marks)**

How many cat breeds were captured in this dataset?

In [9]:
def total_cat_breeds():
    query = """ 
    select 
        count(*) as TOTAL_CAT_BREED
    from(
    select 
        BREED
    from liscensed_pets 
    where PET_TYPE = 'Cat'
    group by BREED) as t;
    """

    return pd.read_sql(sql = query, con = engine)

print(total_cat_breeds())

   TOTAL_CAT_BREED
0               67


There are 67 cat breeds

**(B3): (2 marks)**

Which dog breed is licensed the most? (1 Mark) How many of these have a license date in 2023? (1 Mark)

In [10]:
# return the name of dog lisenced most with its number liscensed 
def lisenced_dog_breed():
    query = """ 
    select 
        BREED, 
        count(COUNT) as lISENCED_COUNT
    from liscensed_pets
    where PET_TYPE = 'Dog'
    group by BREED
    order by count(COUNT) DESC
    limit 1;
    """

    return(pd.read_sql(sql= query, con = engine))

# return number of dog lisenced in 2023
def lisenced_dog_breed2023():
    query = """ 
    select 
        BREED, 
        count(COUNT) as lISENCED_COUNT
    from liscensed_pets
    where PET_TYPE = 'Dog'
    AND YEAR = 2023
    group by BREED
    order by count(COUNT) DESC
    limit 1;
    """

    return(pd.read_sql(sql= query, con = engine))

print("most lisenced dog:")
print(lisenced_dog_breed())

print("Its number of lisenced in 2023:")
print(lisenced_dog_breed2023())

most lisenced dog:
                BREED  lISENCED_COUNT
0  LABRADOR RETRIEVER            5033
Its number of lisenced in 2023:
                BREED  lISENCED_COUNT
0  LABRADOR RETRIEVER             164


The most liscensed dog breed is LABRADOR RETRIEVER and there are 164 of them lisenced in 2023

**(B4): (2 marks)**

Create a table listing how many animals of breed types 'RAGDOLL', 'BEAGLE' have been licensed in each neighbourhood.

In [11]:
def num_liscensed_neighborhood():
    # create a table if not exist
    query1 = text("""
    CREATE TABLE IF NOT EXISTS lisenced_count_neighborhood(
        NEIGHBOURHOOD TEXT DEFAULT NULL,
        BREED TEXT DEFAULT NULL,
        PET_TYPE TEXT DEFAULT NULL,
        NUMBER_LISENCED INT DEFAULT NULL
    );""")
    # truncate before inserting information, avoid duplicate information 
    query2 = text(""" Truncate table lisenced_count_neighborhood;
    """)
    # insert data into table 
    query3 = text("""
    INSERT INTO lisenced_count_neighborhood
    SELECT
        NEIGHBOURHOOD,
        BREED,
        PET_TYPE,
        COUNT(*) AS NUMBER_LISENCED
    FROM liscensed_pets
    WHERE BREED = 'BEAGLE' OR BREED = 'RAGDOLL'
    GROUP BY NEIGHBOURHOOD, BREED;
    """)
    # excute all queries as one query
    with engine.connect() as conn:
        conn.execute(query1)
        conn.execute(query2)
        conn.execute(query3)
        conn.commit()

    query = text("""
    Select *
    from lisenced_count_neighborhood
    limit 5
    """)

    return pd.read_sql(query, con = engine)

    

print(num_liscensed_neighborhood())

    NEIGHBOURHOOD    BREED PET_TYPE  NUMBER_LISENCED
0    Abbottsfield   BEAGLE      Dog                1
1    Abbottsfield  RAGDOLL      Cat                1
2          Albany  RAGDOLL      Cat                5
3  Alberta Avenue   BEAGLE      Dog                3
4  Alberta Avenue  RAGDOLL      Cat                4


**(B5): (2 marks)**

What is the ratio of cats to dogs in this dataset?

In [12]:
def ratio():
    # count(*) counting the number for cat and dog
    # lag(count(*)) make a delay on the column of count 
    # the ratio is the ratio of 
    query = text("""
    select 
        PET_TYPE,
        CONCAT(COALESCE(LAG(PET_TYPE) OVER(ORDER BY PET_TYPE), 'NULL'), '/', PET_TYPE) AS ANIMAL_ORDER,
        coalesce(lag(Count(*)) over (order by count(*))) / Count(*) as RATIO
    from liscensed_pets
    where PET_TYPE in ('Dog', 'Cat')
    group by PET_TYPE
    """)
    return_df = pd.read_sql(sql = query, con = engine)
    return return_df
print(ratio())

  PET_TYPE ANIMAL_ORDER   RATIO
0      Cat     NULL/Cat     NaN
1      Dog      Cat/Dog  0.4965


The ratio of cat over dog (num(cat)/ num(dog)) is 0.4965

## Part C: Detailed analysis (20 marks)

Now consider being given a task to make sense of the distribution of licensed pets based on this dataset.

**(C1):(4 marks)**

Create two guiding questions to use in your analysis, and include them below as Markdown. As a starting point (and remember you are not limited to only these!), you may want to consider the following ideas:
- Focus on a specific set of data from the dataset that interests you.
- Consider what you know from your study in lecture of a similar dataset from the City of Calgary. What is the same, and what is different?
- What kinds of things might a prospective pet owner need to consider?

1) 

In 2024, Betwen Calgary and Edmoton, which cities have more number of lisenced pet based on each types of pet. 

2) 
What is the top 5 neighborhood that have lowest Ratio of spayed or neutered liscensed pet (SPAYED_OR_NEUTERED / total), neglect the one that have less than ten record of lisences or no record for actual SPAYED_OR_NEUTERED?


**(C2): (12 marks)** 

Write at least four queries (that is, two queries for each question) which you believe will address one of your guiding questions. Clearly indicate which queries address your questions. You may wish to include a comment to explain why this query will help address your question.

1) In 2024, Betwen Calgary and Edmoton, which cities have more number of lisenced pet based on each types of pet. 

In [13]:
# The link for the csv file is in the reference
pet_calgary = pd.read_csv('Licensed_Pets.csv')
pet_calgary['DATE'] = pd.to_datetime(pet_calgary['DATE'], format = '%Y %B')

# rename columns for for callable in sql
pet_calgary['LICENSE_VOLUME'] = pet_calgary['LICENSE VOLUME']
pet_calgary['COMMUNITY_CODE'] = pet_calgary['COMMUNITY CODE']
pet_calgary['COMMUNITY_NAME'] = pet_calgary['COMMUNITY NAME']
pet_calgary['ROW_ID'] = pet_calgary['ROW ID']

# delete ',' quotation in Lisenced_VOLUME and change type to numeric 
pet_calgary['LICENSE_VOLUME'] = pd.to_numeric(pet_calgary['LICENSE_VOLUME'].str.replace(',',''))
for i in ['LICENSE VOLUME', 'COMMUNITY CODE','COMMUNITY NAME', 'ROW ID']:

    if i in pet_calgary.columns:
        pet_calgary = pet_calgary.drop(columns=[i], inplace= False)



# create Calgary_pet_lisenced table in the database
def create_Calgary_pet_table():
    pet_calgary.to_sql(        
        name = 'Calgary_pet_lisenced',
        con = engine,
        if_exists= 'replace',
        index= False)
    return None

# call the function to import the table in database
# create_Calgary_pet_table()

In [14]:
def lisenced_compared_2024():
    # 3 main query
    # main querry for joining 2 tables with key is petype
    # first subquerry for finding number lisences based on pet type in edmotion
    # second subquery for finding number lisences based on pet type in Calgary
    # the reuslt is restored in column CITIES_LARGER_LISCENCES for city that have larger amount lisences
    query = text("""
    select
        ep.PET_TYPE as PET_TYPE,
        ep.EDMOTON_LISENCED_COUNT, 
        Coalesce(cp.CALGARY_LISENCED_COUNT,0) as CALGARY_LISENCED_COUNT ,
        Case 
            when ep.EDMOTON_LISENCED_COUNT > coalesce(CALGARY_LISENCED_COUNT,0) then 'Edmoton'
            when ep.EDMOTON_LISENCED_COUNT < coalesce(CALGARY_LISENCED_COUNT,0) then 'Calgary'
            else 'equal'
        END as CITIES_LARGER_LISCENCES
    
    from (
    Select
        PET_TYPE,
        sum(COUNT) as EDMOTON_LISENCED_COUNT
    from liscensed_pets
    where YEAR = 2024
    group by PET_TYPE) as ep

    left join (
    Select 
        Case 
            when ANIMAL = 'CATS' then 'Cat'
            when ANIMAL = 'DOGS' then 'Dog'
        ENd as PET_TYPE,
        sum(LICENSE_VOLUME) as CALGARY_LISENCED_COUNT
    from Calgary_pet_lisenced
    where YEAR(DATE) = 2024
    group by ANIMAL) as cp

    on ep.PET_TYPE = cp.PET_TYPE""")

    return pd.read_sql(sql = query, con = engine) # return as a dataframe

print(lisenced_compared_2024())


  PET_TYPE  EDMOTON_LISENCED_COUNT  CALGARY_LISENCED_COUNT  \
0      Cat                  7478.0                339647.0   
1      Dog                 15197.0                789346.0   
2  Pigeons                    32.0                     0.0   

  CITIES_LARGER_LISCENCES  
0                 Calgary  
1                 Calgary  
2                 Edmoton  


From the table, it can be said that, in 2024, Calgary have more liscensd Dog and Cat compared to Edmoton but Edmoton have more lisenced Pigeons compared to Calgary. The reasons for the high difference in the data between Edmoton and Calgary raise from the fact that Calgary store the data through many year will edmoton only stored the data in that specific year. 

2) what is the top 5 neighborhood that have lowest Ratio of spayed or neutered liscensed pet (SPAYED_OR_NEUTERED / total), neglect the one that have less than ten record of lisences or no record for actual SPAYED_OR_NEUTERED?

In [15]:
def ratio_SPAYED_OR_NEUTERED():

    # three query 1 main and 2 subquery
    # main query joining 2 table and store the ratio value in RATIO_SPAYED_OR_NEUTERED column

    # first subquery find the number of MIX_SPAYED_OR_NEUTERED in edmoton
    # MIX_SPAYED_OR_NEUTERED is the combination of spayed_or_neutered and not spayed_or_neutered

    # second subquery find the number of TRUE_SPAYED_OR_NEUTERED in edmoton 
    # TRUE_SPAYED_OR_NEUTERED is the number of actually spayed_or_neutered

    # the one does not have record for TRUE_SPAYED_OR_NEUTERED will be assigned as 0 in ratio 

    # here I will eliminate the neighbor that have ratio as 1 and 0 since in these 2 neighbor
    # usualy there is no record for TRUE_SPAYED_OR_NEUTERED(ratio = 0) 
    # and too less amount of record for liscensed pet(ratio =1)
    
    query = text("""
    select 
        mn.NEIGHBOURHOOD,
        coalesce(tsn.TRUE_SPAYED_OR_NEUTERED / mn.MIX_SPAYED_OR_NEUTERED,0) as RATIO_SPAYED_OR_NEUTERED,
        COALESCE(tsn.TRUE_SPAYED_OR_NEUTERED  / mn.MIX_SPAYED_OR_NEUTERED, 0) 
        - AVG(COALESCE(tsn.TRUE_SPAYED_OR_NEUTERED / mn.MIX_SPAYED_OR_NEUTERED, 0)) 
        OVER () AS MEAN_DIFF,
        rank() over (order by coalesce(tsn.TRUE_SPAYED_OR_NEUTERED / mn.MIX_SPAYED_OR_NEUTERED,0) ASC) as RANK
    from
    (select
        NEIGHBOURHOOD,
        COUNT(SPAYED_OR_NEUTERED) as MIX_SPAYED_OR_NEUTERED
    from liscensed_pets 
    group by NEIGHBOURHOOD)as mn

    left join(
    select 
        NEIGHBOURHOOD,
        count(SPAYED_OR_NEUTERED) as TRUE_SPAYED_OR_NEUTERED
    from liscensed_pets
    where lower(SPAYED_OR_NEUTERED) = 'yes'
    group by NEIGHBOURHOOD
    ) as tsn
                 
    on mn.NEIGHBOURHOOD = tsn.NEIGHBOURHOOD
                 
    where coalesce(tsn.TRUE_SPAYED_OR_NEUTERED / mn.MIX_SPAYED_OR_NEUTERED,0) != 0
    and coalesce(tsn.TRUE_SPAYED_OR_NEUTERED / mn.MIX_SPAYED_OR_NEUTERED,0) != 1
                 
    Limit 5
    """)

    return pd.read_sql(sql = query, con = engine)

print(ratio_SPAYED_OR_NEUTERED())


                         NEIGHBOURHOOD  RATIO_SPAYED_OR_NEUTERED  MEAN_DIFF  \
0                 Armstrong Industrial                    0.0323  -0.841336   
1  Edmonton Energy And Technology Park                    0.2727  -0.600936   
2                  Mistatim Industrial                    0.3077  -0.565936   
3      Winterburn Industrial Area West                    0.3333  -0.540336   
4               Bonaventure Industrial                    0.5000  -0.373636   

   RANK  
0     1  
1     2  
2     3  
3     4  
4     5  


The top 5 Neighborhood that have lowest RATIO_SPAYED_OR_NEUTERED are Armstrong Industrial, Edmonton Energy And Technology Park, Mistatim Industrial, Winterburn Industrial Area West, Bonaventure Industrial. These areas are industrial areas rather than redidential zone in edmoton, thus they will have fewer resident properties, leading to lower pet care practice available. Thus pets here will have lower chance for actually be SPAYED_OR_NEUTERED. Especially in Armstrong Industrial, with the ratio is 0.0323, which is almost not SPAYED_OR_NEUTERED, this come from the fact that they are big industrial company so animal is rather kept for working rather than pets. 

**(C3): (4 marks)**

Use the Markdown cell below. What kind of data would be interesting to have to be able to make more sense of the dataset? (1 mark) 

Data for population in those region in edmoton should be a great addition for the question2 

Use the Markdown cell below. Discuss how you could use this additional information to extend one of your guiding questions. (3 Marks)

By using this data, I can find out the does the population of each region does affect the ratio of
SPAYED_OR_NEUTERED, or it affected by other factors such as awareness, group og human, for 
example: students, workers,.... This can be done by joining 2 dataset using Neighborhood as global 
key then see the ratio of RATIO_SPAYED_OR_NEUTERED and number of RATIO_OWN_PET_POPULATION in that 
region

## Part D: Reflection (9 marks)

1. You may have been required to complete a number of tasks to ensure that you were working with the provided dataset in your database the way you wanted. Which task did you find the most difficult to accomplish? (3 marks)





Task B4 is most challenged for me compared another one.

2. Why was that? Were there aspects that were easier, or did you find this task to be uniformly difficult? (3 marks)

The reason is that  it require one to manage the data type of the table, it should be reasonable 
in order to make futher analsing at the use domain. Besides, you also have to make sure the table 
is actually keep up with the real time processing, for example, how the table get refresh when 
ever the query is run, for example: update new rows or clear the whole table to insert whole new 
values. 

3. Which of the provided resources discussed in class or provided as part of the course reading list were the most useful? What did you like most about this resource? (3 marks)

The most hellful resource is the dataset of Calgary pet liscensed, this help me to make a comparsion between two cities and see the preference in different kind of pet. Futhermore, it can extend to others factors that relate specific to two villes such as economies, population, etc....

In [16]:
# Use this cell to include some code to dispose of your SQLAlchemy engine object
engine.dispose()

## References

City of Edmonton (2025). Retrieved From: https://data.edmonton.ca/Community-Services/Pet-Licenses-by-Neighbourhood/5squ-mg4w. 

Licensed Pets | Open Calgary, Retrieved From: https://data.calgary.ca/Services-and-Amenities/Licensed-Pets/5dgy-88cq/data_preview.

